Opam init complains that sandboxing isn't working on platform macports

Hi,

I'm trying to use the opam 2.1 beta version (I tried both the published beta, and a version compiled from the current trunk) with a clean install. When I run `opam init` I see an error message about sandboxing not working.

```
% opam init --bare
[NOTE] Will configure from built-in defaults.
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.

[ERROR] Sandboxing is not working on your platform macports:
        "~/.opam/opam-init/hooks/sandbox.sh build sh -c echo SUCCESS >/tmp/t && cat /tmp/t" exited with code 1
        "/Users/asoni/.opam/opam-init/hooks/sandbox.sh: line 61: OPAM_SWITCH_PREFIX: unbound variable"
Do you want to disable it?  Note that this will result in less secure package builds, so please ensure that you have some other isolation mechanisms in place
(such as running within a container or virtual machine). [y/N]
```

If I ignore the warning and continue by pressing `N` the install seems to succeed without issues. (opam doesn't automatically source the correct switch for me anymore even though the shell hooks have been installed, but the install overall seems to work fine). 

I am using macOS 10.15 (Catalina) along with macports. This error wasn't present in the opam 2.0 series. 
I'm wondering if this is an environment variable issue and I need to setup something to let Opam know about macports? It seems like the new depext integration is also having trouble finding dependencies like libffi via macports. It complains about them not being installed even though they are.

--- a/master_changes.md
+++ b/master_changes.md
@@ -4,13 +4,13 @@ Possibly scripts breaking changes are prefixed with ✘.
 New option/command/subcommand are prefixed with ◈.
 
 ## Version
-  * Bump version to '2.1.0~beta3' [@altgr]
+  * Bump version to '2.1.0~beta3' [#4351 @AltGr]
 
 ## Global CLI
   *
 
 ## Init
-  *
+  * Fix sandbox check with not yet set opam environment variables [#4370 @rjbou - fix #4368]
 
 ## Config Upgrade
   *

--- a/src/client/opamAuxCommands.ml
+++ b/src/client/opamAuxCommands.ml
@@ -492,8 +492,11 @@ let check_and_revert_sandboxing root config =
   | cmd::_ ->
     let test_cmd = [ "sh"; "-c"; "echo SUCCESS >/tmp/t && cat /tmp/t" ] in
     let working_or_noop =
+      let env =
+        Array.append [| "OPAM_SWITCH_PREFIX=/dev/null" |] (Unix.environment ())
+      in
       try
-        OpamSystem.read_command_output ~allow_stdin:false (cmd @ test_cmd)
+        OpamSystem.read_command_output ~env ~allow_stdin:false (cmd @ test_cmd)
         = ["SUCCESS"]
       with e ->
         (OpamConsole.error "Sandboxing is not working on your platform%s:\n%s"

--- a/src/state/shellscripts/bwrap.sh
+++ b/src/state/shellscripts/bwrap.sh
@@ -85,6 +85,9 @@ add_dune_cache_mount() {
   add_mount rw "$u_dune_cache" "$dune_cache"
 }
 
+# When using opam variable that must be defined at action time, add them also
+# at init check in OpamAuxCommands.check_and_revert_sandboxing (like
+# OPAM_SWITCH_PREFIX).
 # This case-switch should remain identical between the different sandbox implems
 COMMAND="$1"; shift
 case "$COMMAND" in

--- a/src/state/shellscripts/sandbox_exec.sh
+++ b/src/state/shellscripts/sandbox_exec.sh
@@ -54,6 +54,9 @@ add_dune_cache_mount() {
   add_mounts rw "$DUNE_CACHE"
  }
 
+# When using opam variable that must be defined at action time, add them also
+# at init check in OpamAuxCommands.check_and_revert_sandboxing (like
+# OPAM_SWITCH_PREFIX).
 # This case-switch should remain identical between the different sandbox implems
 COMMAND="$1"; shift
 case "$COMMAND" in

